[DOCS][8.x] Add `Using ESQL for search` landing page

leemthompo opened the pull request:
One-stop shop for search with ESQL


## TODO

- [ ] Add/uncomment links to [tutorial](https://docs-v3-preview.elastic.dev/elastic/docs-content/tree/main/solutions/search/esql-full-text-filter-tutorial) once backported to 8.x
- [x] Does KQL belong here???


## Screenshot

![image](https://github.com/user-attachments/assets/02878952-34df-4498-b746-9b6f31c44dc8)

leemthompo committed:
[DOCS][8.x] Add Search for ESQL landing page

leemthompo committed:
tweak wording

carlosdelest commented on docs/reference/esql/esql-for-search.asciidoc:
```diff
@@ -0,0 +1,191 @@
+
+[[esql-for-search-full-text]]
+==== Full text search
+
```
We may introduce why full text functions are preferrable to use to match text than other ESQL constructs:

- Performance
- Analyzers
- Multi valued texts

[This blog](https://www.elastic.co/search-labs/blog/filtering-in-esql-full-text-search-match-qstr) explored some of these

leemthompo replied:
I'll add this mention, I've linked to blog in **Learn more** section but agree would be good to surface this info here too 👍

carlosdelest commented on docs/reference/esql/esql-for-search.asciidoc:
```diff
@@ -0,0 +1,191 @@
+----
+
+For complete details, refer to the <<query-dsl-query-string-query, Query DSL `query_string` docs>>.
+
```
I wonder if we should move KQL up here - it's another search function, even if it has its specific migration use case. 🤔

carlosdelest commented on docs/reference/esql/esql-for-search.asciidoc:
```diff
@@ -0,0 +1,191 @@
+[[esql-for-search-scoring]]
+==== Relevance scoring
+
+To get the most relevant results first, you need to retrieve and sort by relevance score:
```
Should we separate the filtering vs searching use cases? It might be useful for users to think about just removing data and sorting by something else, or sort by actual relevance.

carlosdelest commented on docs/reference/esql/esql-for-search.asciidoc:
```diff
@@ -0,0 +1,191 @@
+
+* If you don't include `METADATA _score` in your query, this only performs filtering operations with no relevance calculation.
+* When you include `METADATA _score`, search functions included in `WHERE` conditions contribute to the relevance score.
+* Filtering operations (like range conditions and exact matches) don't affect the score.
```
```suggestion
* Filtering operations that are not search functions, like range conditions and exact matches, don't affect the score.
```

carlosdelest commented on docs/reference/esql/esql-for-search.asciidoc:
```diff
@@ -0,0 +1,191 @@
+When working with relevance scoring in ES|QL, it's important to understand how `_score` works:
+
+* If you don't include `METADATA _score` in your query, this only performs filtering operations with no relevance calculation.
+* When you include `METADATA _score`, search functions included in `WHERE` conditions contribute to the relevance score.
```
Clarifying more what impacts score. Let's make it hard for users not to understand this 😉 

```suggestion
* When you include `METADATA _score`, any search function included in `WHERE` conditions contribute to the relevance score. This means that every occurrence of `MATCH`, `QSTR` and `KQL` will affect the score.
```

leemthompo replied:
💯

carlosdelest reviewed (commented):
Looks 🔝

leemthompo commented on docs/reference/esql/esql-for-search.asciidoc:
```diff
@@ -0,0 +1,191 @@
+===== Background concepts
+
+* <<analysis>>: How text is processed for full-text search
+* <<semantic-search>>: Overview of semantic search in Elasticsearch
```
good idea!

leemthompo committed:
Revamp per Carlos' review

leemthompo committed:
fix typo

leemthompo committed:
make heading consistent

carlosdelest reviewed (approved):
This looks 💯  to me.

carlosdelest commented on docs/reference/esql/esql-for-search.asciidoc:
```diff
@@ -0,0 +1,211 @@
+
+* **Filtering** removes non-matching documents without calculating relevance scores
+* **Searching** both filters documents and ranks them by how well they match the query
+
```
I wonder if we should hint at the performance. Search will be less performant than filtering, as scores need to be calculated. 🤔

leemthompo committed:
mention filtering is faster than searching

leemthompo commented on docs/reference/esql/esql-for-search.asciidoc:
```diff
@@ -0,0 +1,212 @@
+* **Searching** both filters documents and ranks them by how well they match the query
+
+Note that filtering is faster than searching, because it doesn't require score calculations.
+[[esql-for-search-scoring]]
```
```suggestion

[[esql-for-search-scoring]]
```

leemthompo committed:
add space

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

ioanatia reviewed (approved):
looks great - is it worth adding a note somewhere saying these are tech preview features?

leemthompo commented:
> is it worth adding a note somewhere saying these are tech preview features?

yes! thanks for catching :)

leemthompo committed:
Add tech preview banner

leemthompo committed:
move to "Using ESQL" section with improved title

Renames "Search with ESQL" to "Using ESQL for search" and relocates
the page from language reference to usage section for better discoverability.

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.18](https://github.com/elastic/elasticsearch/pull/125653)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/125653">](https://github.com/elastic/elasticsearch/pull/125653) |
